### PR TITLE
POC for digit as first char in baseName property value.

### PR DESCRIPTION
**Pull request by bd82**

### CODE REVIEW

**Comment by bd82** [comment]
Discussion in #241

File: lib/dsl/ast_builder.js [hunk 1]
```diff
@@ -482,6 +482,8 @@ class JDLAstBuilderVisitor extends BaseJDLCSTVisitor {
       return stringImage.substring(1, stringImage.length - 1);
     } if (ctx.BOOLEAN) {
       return ctx.BOOLEAN[0].image === 'true';
+    } if (ctx.NAME_CAN_START_DIGIT) {
```
Comment (MathieuAA) [inline lib/dsl/ast_builder.js]: Didn't notice what was previously there, but each if should start on its own line.

File: lib/dsl/jdl_parser.js [hunk 2]
```diff
@@ -571,6 +571,7 @@ module.exports = class JDLParser extends Parser {
   configValue() {
     this.RULE('configValue', () => {
       this.OR([
+        { ALT: () => this.CONSUME(LexerTokens.NAME_CAN_START_DIGIT) },
```
Comment (MathieuAA) [inline lib/dsl/jdl_parser.js]: I don't like it at all. There's no control over each property, we just say "okay, these types are possible and let the validator check everything's okay".

File: lib/dsl/lexer.js [hunk 3]
```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
 
 const tokens = {};
 
+const baseNamePattern = 'baseName';
```
Comment (MathieuAA) [inline lib/dsl/lexer.js]: It's not really a pattern... its use is, but it's not. The name is kinda misleading.

File: lib/dsl/lexer.js [hunk 4]
```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
...
 const namePattern = /[a-zA-Z_][a-zA-Z_\-\d]*/;
-// All the names tokens of the pegjs implementation have been merged into
+// using start anchor because we are matching with this regexp directly in a custom pattern.
+const namePatternWithStartDigit = /^[\da-zA-Z_][a-zA-Z_\-\d]*/;
```
Comment (MathieuAA) [inline lib/dsl/lexer.js]: Some rules should be enforced here, whether to use `\d` or `0-9`. Having both is not that pretty. By the way, why isn't it ` const namePatternWithStartDigit = /^[\da-zA-Z_][a-zA-Z_\-\d]*$/; ` the `$` sign marking the end of the local matching)

File: lib/dsl/lexer.js [hunk 5]
```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
...
+/**
+ * Accepts an array of array representing character ranges
+ * e.g: [['a', 'd'], ['0', '2']]
+ *
+ * And returns a set of the "expanded" ranges in charCodes values.
+ * e.g: [97, 99, 98 , 100, 48, 49, 50]
+ *   - 'a'.charCodeAt(0) === 97
+ *   - 'a'.charCodeAt(0) === 100
+ */
+function getRanges(charRanges) {
```
Comment (MathieuAA) [inline lib/dsl/lexer.js]: nice idea

File: lib/dsl/lexer.js [hunk 6]
```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
...
+
+    ranges = ranges.concat(_.range(startCode, endCode));
+  });
+
+  return _.uniq(ranges);
+}
+
+// Special NAME Token which follows a baseName config value.
+// It allows starting with numbers or even being only digits.
+const NAME_CAN_START_DIGIT = chevrotain.createToken({
```
Comment (MathieuAA) [inline lib/dsl/lexer.js]: some wording to do, "can start with digit", nothing critical but I get the meaning

File: lib/dsl/lexer.js [hunk 7]
```diff
@@ -23,8 +23,51 @@ const Lexer = chevrotain.Lexer;
...
+  pattern: (text, offset, matchedTokens) => {
+    const lastToken = _.last(matchedTokens);
+    if (lastToken && lastToken.image === baseNamePattern) {
+      return namePatternWithStartDigit.exec(text.substring(offset));
+    }
+
+    return null;
+  },
+  // this optional property enables performance optimizations in the Lexer.
+  start_chars_hint: getRanges([['0', '9'], ['a', 'z'], ['A', 'Z'], ['_', '_']]),
```
Comment (MathieuAA) [inline lib/dsl/lexer.js]: If there are inputs like `['_', '_']`, maybe not iterating over it is better by passing a string and checking with `Array::isArray`

File: lib/dsl/jdl_parser.js (hunk 2 above)
Comment (bd82) [reply lib/dsl/jdl_parser.js]: That is basic separation of concerns, Syntax versus Type Checks.
Also implemented like  this in the JDL IDE.
https://github.com/jhipster/jhipster-ide/blob/master/jdl/io.github.jhipster.jdl/src/io/github/jhipster/jdl/JDL.xtext#L74

Its like applying a JSON schema on a JSON input.
You can expand it to the full grammar with the very specific types for each config key.
There is nothing preventing doing so. 

A gentle reminder that you yourself complained that maintaining the old system was a chore.
https://github.com/jhipster/jhipster-core/issues/198#issuecomment-374896968
This way its much easier to add new config keys (you don't even have to change the parser!).

File: lib/dsl/lexer.js (hunk 3 above)
Comment (bd82) [reply lib/dsl/lexer.js]: Yes I see your point

File: lib/dsl/lexer.js (hunk 4 above)
Comment (bd82) [reply lib/dsl/lexer.js]: I agree.

File: lib/dsl/lexer.js (hunk 6 above)
Comment (bd82) [reply lib/dsl/lexer.js]: Will fix if we move ahead with this change.

File: lib/dsl/lexer.js (hunk 7 above)
Comment (bd82) [reply lib/dsl/lexer.js]: I don't understand the comment can you demonstrate?

File: lib/dsl/ast_builder.js (hunk 1 above)
Comment (bd82) [reply lib/dsl/ast_builder.js]: It was a series of "else if" which you modified to plain ifs in this commit
https://github.com/jhipster/jhipster-core/commit/fe6614892f03cf02b9baf84a17546ea93140a605#diff-9fa3402fa25eaa022ea233eb8dbb86c9
Which is why the formatting is not valid...

But style X or style Y are irrelevant, the main issue is that formatting is not done **automatically.** Basically we(Humans) should not invest time in doing something a computer would do better and more consistently (prettier/standardjs).

Particularly as we all have day jobs and limited time to invest in OSS. 😭

Comment (deepu105) [reply lib/dsl/ast_builder.js]: I'm always ok to add prettier. I thought it was already done here??

Comment (deepu105) [reply lib/dsl/ast_builder.js]: I'm sorry just found the old ticket and realized I was being ignorant there, I think aging a bit more has made me a bit wiser :wink: 
https://github.com/jhipster/jhipster-core/issues/158

Comment (bd82) [reply lib/dsl/ast_builder.js]: np, its good to use this example use case to re-raise the question of standard consistent formatting tool. 👍

File: lib/dsl/jdl_parser.js (hunk 2 above)
Comment (MathieuAA) [reply lib/dsl/jdl_parser.js]: Well as Deepu said, aging made me a bit wiser ;)

File: lib/dsl/lexer.js (hunk 7 above)
Comment (MathieuAA) [reply lib/dsl/lexer.js]:
```javascript
getRanges([
  ['0', '9'],
  ['a', 'z'],
  ['A', 'Z'],
  '_'
]);
```

Comment (bd82) [reply lib/dsl/lexer.js]: Yes this version is nicer.

**Pull request closed without merging**